tianshou/policy/imitation/bcq.py

```diff
@@ -0,0 +1,326 @@
+    :param torch.optim.Optimizer actor_optim: the optimizer for actor network.
+    :param torch.nn.Module critic1: the first critic network. (s, a -> Q(s, a))
+    :param torch.optim.Optimizer critic1_optim: the optimizer for the first
+        critic network.
+    :param torch.nn.Module critic2: the second critic network. (s, a -> Q(s, a))
+    :param torch.optim.Optimizer critic2_optim: the optimizer for the second
+        critic network.
+    :param VAE vae: the VAE network, generating actions similar
+        to those in batch. (s, a -> generated a)
+    :param torch.optim.Optimizer vae_optim: the optimizer for the VAE network.
+    :param Union[str, torch.device] device: which device to create this model on.
+        Default to cpu.
+    :param float gamma: discount factor, in [0, 1]. Default to 0.99.
+    :param float tau: param for soft update of the target network.
+        Default to 0.005.
+    :param float lmbda: param for Clipped Double Q-learning. Default to 0.75.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.BasePolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        actor: Perturbation,
+        actor_optim: torch.optim.Optimizer,
+        critic1: torch.nn.Module,
+        critic1_optim: torch.optim.Optimizer,
+        critic2: torch.nn.Module,
+        critic2_optim: torch.optim.Optimizer,
+        vae: VAE,
+        vae_optim: torch.optim.Optimizer,
+        device: Optional[Union[str, torch.device]] = "cpu",
+        gamma: float = 0.99,
+        tau: float = 0.005,
+        lmbda: float = 0.75,
+        **kwargs: Any
+    ) -> None:
+        # actor is Perturbation!
+        super().__init__(**kwargs)
+        self.actor = actor
+        self.actor_target = copy.deepcopy(self.actor)
+        self.actor_optim = actor_optim
+
+        self.critic1 = critic1
+        self.critic1_target = copy.deepcopy(self.critic1)
+        self.critic1_optim = critic1_optim
+
+        self.critic2 = critic2
+        self.critic2_target = copy.deepcopy(self.critic2)
+        self.critic2_optim = critic2_optim
+
+        self.vae = vae
+        self.vae_optim = vae_optim
+
+        self.gamma = gamma
+        self.tau = tau
+        self.lmbda = lmbda
+        self.device = device
+
+    def train(self, mode: bool = True) -> "BCQPolicy":
+        """Set the module in training mode, except for the target network."""
+        self.training = mode
+        self.actor.train(mode)
+        self.critic1.train(mode)
+        self.critic2.train(mode)
+        return self
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        """Compute action over the given batch data."""
+        # state: None, input: "obs"
+        # There is "obs" in the Batch
+        # obs: 10 groups. Each group has a state. shape: (10, state_dim)
```

Comment:
why 10? Can it be another number?

Reply:
I forgot to modify it when writing this code. I think it is the training_num or test_num. I'll modify it.